### PR TITLE
MAINT: handle not implemented arguments centrally, via a dedicated annotation

### PR DESCRIPTION
**Pull request by ev-br**
as discussed in https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/112#discussion_r1162861446

This does seem to gain a marginal LOC reduction and some possible readability improvement. Two points:

- I'm {ab,re}using the `NotImplemented` singleton as the annotation. Introducing a new `NotImplementedType` is a bit awkward since `typing.NotImplementedType` is a thing from python 3.10, and we need to cover 3.8+. A simple alternative is to just use `type(NotImplemented)` instead of `NotImplemented`. Don't have a firm opinion, am open to suggestions.

- NumPy uses a weird `_NoValue` sentinel in some cases; we mimicked it by defining an alias `NoValue = None`. In our usage we typically raise NotImplementedError on things with the `NoValue` default. Now that we have a dedicated annotation, `NoValue` alias seems extraneous. Anyone sees a need for it?

### CODE REVIEW

File: torch_np/_ufuncs.py [hunk 1]
```diff
@@ -6,13 +6,10 @@
 
 from . import _binary_ufuncs_impl, _helpers, _unary_ufuncs_impl
 from ._detail import _dtypes_impl, _util
-from ._normalizations import ArrayLike, DTypeLike, OutArray, SubokLike, normalizer
+from ._normalizations import ArrayLike, DTypeLike, OutArray, normalizer
 
 
 def _ufunc_preprocess(tensors, where, casting, order, dtype, subok, signature, extobj):
```
Comment (lezcano) [inline torch_np/_ufuncs.py:12]: nb. It'd be better to use the `normalizer` decorator directly when defining the `@wrapped` function, to be more explicit that they're being normalized. It's happened to me a couple times that I forgot that these functions are normalized just find that they are being normalized after calling `decor_{unary,binary}`.

**Review by lezcano — COMMENTED** [review]
I really like this! It really shows that, either you should be processing an arg in the implementation, or you should mark it as `NotImplemented`, so there's no ambiguity for "did you forget to use this arg, or is it just not implemented?"

About the `_NoValue`, it makes sense to remove it and simply replace it by `None`.

About the `NotImplementedType`, I'd say it's best to have the usual pattern of

```python
try:
  from typing import NotImplementedType
except ImportError:
  NotImplementedType = typing.TypeVar("NotImplementedType")
```
The reason is that `NotImplemented` is an object, not a class, so using it for typing is a bit odd. WDYT?

File: torch_np/_ufuncs.py (hunk 1 above)
Comment (ev-br) [reply torch_np/_ufuncs.py:18]:
```
def decorator(...):
    @another_decorator
    def wrapped(...):
        ....
```

"I put a decorator inside your decorator", wow :-). Readability counts, so done in the updated PR.

**Review by lezcano — APPROVED** [review]
Great, thank you!

**Pull request merged**